enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,467 @@
+the implementation should implement a way for administrators of an OpenShift
+cluster to configure the specifics of how and when scaling will occur.
+
+This configuration will be performed using a new custom resource
+`ControlPlaneMachineSetAutoscaling`.
+
+The **required** configurations will include the following properties:
+- `machineConfiguration`: this YAML object copies the [
+  configuration](https://docs.openshift.com/container-platform/4.13/machine_management/control_plane_machine_management/cpmso-configuration.html)
+  from controlplane machinset operator to configure the cloud platform's
+  machinesetup with a machinetype and concrete values that must be adjusted. The
+  implementation should use multiple [discriminated
+  unions](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions).
+  One will specify the `machineType` used:
+  - `machineType`: currently always `machines_v1beta1_machine_openshift_io`, but
+  should be able to support [cluster API](https://cluster-api.sigs.k8s.io/) in
+  the future. 
+    - `machines_v1beta1_machine_openshift_io`: specifies options for the
+    machines API based provisioner.
+      - `machineConfigurations`: specifies the concrete specifications for the
+        machines.
+  
+  The `machineConfiguration` will use a discriminated union again to dispatch
+  for the cloud provider. The dispatch key will be `platform`. To start the
+  supported cloud providers must include:
+  - `Azure` and configuration for `vmSize`
+  - `GCP` and configuration for `machineType`
+  - `AWS` and configuration for `instanceType`
+  - `vSphere` and configuration for `memoryMiB`, `numCPUs` and `diskGiB`.
+  
+  Each element that denotes an instance configuration also has to include a
+  `weight` property to specify the order in which scaling should occur.
+
+The **optional** configurations will include the following properties:
+
+- `syncPeriod`: how often will the operator check resource usage to see if scaling
+  should be performed.
+- `scaleUp`: defines configuration for scaling up. This includes the following
+  sub elements:
+  - `stabilizationWindow`: duration to wait after a scale up before another
+    scaling operation can occur.
+  - `selectPolicy`: `next` to select the next bigger instance size.
+  - `triggerPolicy`: `all` or `any`. Determines if scaling occurs if one or
+    all triggers have to be true.
+  - `triggers`: sets up the list of triggers, which will trigger a scale up. The
+    configuration is based on [custom metrics autoscaling
+    configuration](https://docs.openshift.com/container-platform/4.13/nodes/cma/nodes-cma-autoscaling-custom-trigger.html)
+    and should be kept compatbile to make usage easier for users accustomed to
+    CMA. Specifying multiple triggers, the `triggerPolicy` will determine when a
+    scaleup is performed.
+    An empty list of triggers disables scale up.
```

Comment:
Scaling in the CMA operator seems to be based on the exact values of the metrics returned either by the metrics API trigger or the prometheus trigger (at least from what is documented).
So I guess they aren't facing the problem of loosing information, as the ~ 1 minutes aggregation of the metrics API is all that's used.
I just don't feel comfortable scaling on such a low load average, when our alerts aggregates for 8 hours using the prometheus metric.
I think loosing the data and having to reconstruct it might be ok, given that prometheus could be configured with appropriate metrics if this is not desired.

Reply:
Sorry, I was talking about the configuration of the data within the API resource itself rather than prometheus. To disable the scale up, I have to remove all of my triggers. To re-enable, I have to put the triggers back. I feel like we could probably make that a little more ergonomic right?